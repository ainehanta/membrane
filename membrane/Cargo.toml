[package]
authors = ["Jerel Unruh <mail@unruhdesigns.com>"]
description = "Membrane is an opinionated crate that generates a Dart package from a Rust library. Extremely fast performance with strict typing and zero copy returns over the FFI boundary via bincode."
edition = "2018"
keywords = ["dart", "ffi", "flutter", "bindings", "codegen"]
license = "Apache-2.0"
name = "membrane"
readme = "../README.md"
repository = "https://github.com/jerel/membrane"
version = "0.6.2"

[lib]
crate-type = ["lib"]
doctest = false

[features]
generate = ["membrane_macro/generate"]
skip-generate = ["membrane_macro/skip-generate"]

[dependencies]
allo-isolate = "0.1.11"
bincode = {version = "1.3"}
ffi_helpers = "0.2"
futures = "0.3"
heck = "0.3"
inventory = "0.1"
membrane_macro = {version = "^0.5", path = "../membrane_macro"}
membrane_types = {version = "^0.3", path = "../membrane_types"}
regex = "1.5"
serde = {version = "1.0", features = ["derive"]}
serde-generate = "0.20.4"
serde-reflection = "0.3.5"

[dev-dependencies]
example = {path = "../example"}
pretty_assertions = "1.0.0"
serial_test = "0.5.1"
trybuild = "1.0"
