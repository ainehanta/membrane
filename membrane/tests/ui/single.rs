use futures::Future;
use membrane::{async_dart, sync_dart};

struct JoinHandle {}
impl JoinHandle {
  pub fn abort(&self) {}
}

struct Runtime {}
impl Runtime {
  pub fn spawn<T>(&self, future: T) -> JoinHandle
  where
    T: Future + Send + 'static,
    T::Output: Send + 'static,
  {
    JoinHandle {}
  }
}

static RUNTIME: Runtime = Runtime {};

// attribute errors

#[async_dart]
pub async fn no_namespace() -> i32 {}

#[async_dart(namespace = "a", foo = true)]
pub async fn bad_option() -> i32 {}

#[sync_dart(namespace = "a", os_thread = true)]
pub async fn os_thread_option_used_in_sync_fn() -> i32 {}

// return value errors

#[async_dart(namespace = "a")]
pub async fn no_result() -> i32 {}

#[async_dart(namespace = "a")]
pub async fn bare_tuple() -> Result<(i32, i32), String> {}

#[async_dart(namespace = "a")]
pub async fn top_level_option() -> Option<String> {}

#[async_dart(namespace = "a")]
pub async fn return_fn() -> Result<dyn Fn(), String> {}

#[async_dart(namespace = "a")]
pub async fn option_success() -> Result<Option<i32>, String> {
  Ok(Some(1))
}

#[sync_dart(namespace = "a")]
pub fn emitter_in_sync_return() -> impl membrane::emitter::Emitter<Result<String, String>> {
  membrane::emitter::emitter!()
}

// argument errors

#[async_dart(namespace = "a")]
pub async fn failing_arg(self) -> Result<(), String> {
  Ok(())
}

#[async_dart(namespace = "a")]
pub async fn bad_arg_type(one: i32) -> Result<i32, String> {}

#[async_dart(namespace = "a")]
pub async fn failing_arg_two(foo: &[i8]) -> Result<(), String> {
  Ok(())
}

#[async_dart(namespace = "a")]
pub async fn one_success() -> Result<Vec<i32>, String> {
  Ok(vec![10])
}

fn main() {}
