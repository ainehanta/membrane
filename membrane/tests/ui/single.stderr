error: #[async_dart] expects a `namespace` attribute
  --> tests/ui/single.rs:24:1
   |
24 | #[async_dart]
   | ^^^^^^^^^^^^^
   |
   = note: this error originates in the attribute macro `async_dart` (in Nightly builds, run with -Z macro-backtrace for more info)

error: only `namespace=""`, `disable_logging=true`, `os_thread=true`, and `timeout=1000` are valid options
  --> tests/ui/single.rs:27:1
   |
27 | #[async_dart(namespace = "a", foo = true)]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: this error originates in the attribute macro `async_dart` (in Nightly builds, run with -Z macro-backtrace for more info)

error: `os_thread=true` is not a valid option for `sync_dart`
  --> tests/ui/single.rs:30:1
   |
30 | #[sync_dart(namespace = "a", os_thread = true)]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: this error originates in the attribute macro `sync_dart` (in Nightly builds, run with -Z macro-backtrace for more info)

error: expected enum `Result`
  --> tests/ui/single.rs:36:29
   |
36 | pub async fn no_result() -> i32 {}
   |                             ^^^

error: A tuple may not be returned from an `async_dart` function. If a tuple is needed return a struct containing the tuple.
  --> tests/ui/single.rs:39:37
   |
39 | pub async fn bare_tuple() -> Result<(i32, i32), String> {}
   |                                     ^^^^^^^^^^

error: expected enum `Result`
  --> tests/ui/single.rs:42:36
   |
42 | pub async fn top_level_option() -> Option<String> {}
   |                                    ^^^^^^

error: expected a struct, vec, or scalar type but found `dyn Fn()`
  --> tests/ui/single.rs:45:36
   |
45 | pub async fn return_fn() -> Result<dyn Fn(), String> {}
   |                                    ^^^

error: #[sync_dart] expected a return type of `Result<T, E>` found an emitter
  --> tests/ui/single.rs:53:8
   |
53 | pub fn emitter_in_sync_return() -> impl membrane::emitter::Emitter<Result<String, String>> {
   |        ^^^^^^^^^^^^^^^^^^^^^^

error: not a supported argument type for Dart interop
  --> tests/ui/single.rs:60:26
   |
60 | pub async fn failing_arg(self) -> Result<(), String> {
   |                          ^^^^

error: not a supported argument type for Dart interop, please use i64 instead.
  --> tests/ui/single.rs:65:32
   |
65 | pub async fn bad_arg_type(one: i32) -> Result<i32, String> {}
   |                                ^^^

error: not a supported argument type for Dart interop
  --> tests/ui/single.rs:68:35
   |
68 | pub async fn failing_arg_two(foo: &[i8]) -> Result<(), String> {
   |                                   ^^^^^
